tools/state-viewer/src/apply_chain_range.rs

```diff
@@ -232,9 +233,16 @@ fn apply_block_from_range(
                 return;
             }
         }
+
+        let mut storage =
+            RuntimeStorageConfig::new(*chunk_inner.prev_state_root(), use_flat_storage);
+        if use_trie_for_free {
+            storage.source = StorageDataSource::DbTrieOnly;
+        }
```

Comment:
It seems like use flat storage and source should be exclusive options. Can you structure it that way in code?

Reply:
I have started in such a way, but then I realized you could potentially use both together.. for example use flat storage and still access trie for free. 
No strong opinion if you prefer I'll make the options in CLI exclusive

Reply:
When would it be possible to use both?

Reply:
I think when flat storage is enabled we don't read trie at all (except for the non-inlined values).

Reply:
I'm referring to this struct

```
pub struct RuntimeStorageConfig {
    pub state_root: StateRoot,
    pub use_flat_storage: bool,
    pub source: StorageDataSource,
    pub state_patch: SandboxStatePatch,
}
```

API-wise it gives the impression `use-flat-storage` is orthogonal from the `StorageDataSource`. You can use flat storage or not over a DB that charges gas costs or not. 
Yes not charging gas costs is similar to using flat-storage in some contexts (replaying block)

Reply:
Ah well then this struct is wrong too :) Given the source is already there the 'use_flat_storage' should be derivable from it. I would just remove it. If you don't want to go so far in this PR perhaps you can just add a new ctor? I just dislike this pattern of calling ctor and then immediately overwriting one of the fields, it seems very volatile.

Reply:
Agree, I'll add another constructor!

Reply:
I have made some sweeping changes to make stuff more explicit. No more flags to enable flat storage or free tries, instead now there's an enum `StorageSource` with all different options